File: 2025/typescript/day08.ts
import { type DayModule, type SolutionFactory } from './types'
import * as utils from './utils'

export const part1: Part = input => () => solve('part1', input, 1000)

export const part2: Part = input => () => solve('part2', input)

type Pos = [number, number, number]

type PartId = 'part1' | 'part2'
function solve(
    partId: 'part1',
    input: string[],
    connectionsCount: number,
): number
function solve(partId: 'part2', input: string[]): number
function solve(
    partId: PartId,
    input: string[],
    connectionsCount?: number,
): number {
    const positions: Pos[] = input.map(
        line => line.split(',').map(Number) as Pos,
    )

    const n = positions.length
    const edges: [number, number, number][] = []
    for (let i1 = 0; i1 < n; i1++) {
        const p1 = positions[i1]!
        for (let i2 = i1 + 1; i2 < n; i2++) {
            const p2 = positions[i2]!
            const dx = p1[0] - p2[0]
            const dy = p1[1] - p2[1]
            const dz = p1[2] - p2[2]
            const dist = dx * dx + dy * dy + dz * dz
            edges.push([dist, i1, i2])
        }
    }
    const distances = utils.PriorityQueue.from(edges, (a, b) => a[0] - b[0])

    const uf = new utils.UnionFind(n)
    let stepsLeft = partId === 'part1' ? connectionsCount! : 0
    let lastEdge: [number, number, number] | undefined

    while (true) {
        if (
            (partId === 'part1' && stepsLeft === 0) ||
            (partId === 'part2' && uf.componentCount === 1)
        )
            break

        const currentEdge = distances.pop()
        if (!currentEdge) break
        lastEdge = currentEdge
        const [, i, j] = currentEdge
        if (partId === 'part1') stepsLeft -= 1
        uf.unite(i, j)
    }

    switch (partId) {
        case 'part1': {
            const sizes = uf.sizes()
            const pq = utils.PriorityQueue.from(sizes, (a, b) => b - a)
            return pq.pop()! * pq.pop()! * pq.pop()!
        }
        case 'part2': {
            const [, i, j] = lastEdge!
            return positions[i]![0] * positions[j]![0]
        }
        default:
            utils.unreachable(`Unknown part id: ${partId}`)
    }
}

export const day = import.meta.file.match(/day(\d+)/)![1]!
export const input = await utils.readInputLines(day)
part1.solution = 97384
part2.solution = 9003685096

export const main = import.meta.main
if (main) {
    const module: Module = await import(import.meta.path)

    const testInput = utils.as_lines([
        `
        162,817,812
        57,618,57
        906,360,560
        592,479,940
        352,342,300
        466,668,158
        542,29,236
        431,825,988
        739,650,466
        52,470,668
        216,146,977
        819,987,18
        117,168,530
        805,96,715
        346,949,466
        970,615,88
        941,993,340
        862,61,35
        984,92,344
        425,690,689
`,
    ])

    await utils.tests(
        () =>
            utils.test_all(
                ['Test part 1', 40, () => solve('part1', testInput[0]!, 10)],
                ['Test part 2', 25272, part2(testInput[0]!)],
            ),
        () => utils.test_day(module),
    )
}

type In = typeof input
type Out = number
type Module = DayModule<Out, In>
type Part = SolutionFactory<Out, In>

File: 2025/zig/src/day08.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const aoc = @import("aoc_utils");

pub fn part1(input: []const u8, gpa: Allocator) !Day.Result1 {
    return try solve(.{ .part1 = 1000 }, input, gpa);
}

pub fn part2(input: []const u8, gpa: Allocator) !Day.Result2 {
    return try solve(.part2, input, gpa);
}

const PartId = union(enum) { part1: u16, part2 };
fn Result(partId: PartId) type {
    return switch (partId) {
        .part1 => Day.Result1,
        .part2 => Day.Result2,
    };
}

const Pos = struct { x: u32, y: u32, z: u32 };
const DistanceInfo = struct {
    dist: u64,
    junctionIdx1: usize,
    junctionIdx2: usize,

    fn compare(_: void, a: DistanceInfo, b: DistanceInfo) std.math.Order {
        return std.math.order(a.dist, b.dist);
    }
};

const DistanceQueue = std.PriorityQueue(DistanceInfo, void, DistanceInfo.compare);

const UnionFind = struct {
    parent: []usize,
    rank: []u8,
    component_count: usize,

    fn init(gpa: Allocator, n: usize) !UnionFind {
        const parent = try gpa.alloc(usize, n);
        const rank = try gpa.alloc(u8, n);
        for (0..n) |i| {
            parent[i] = i;
            rank[i] = 0;
        }
        return .{ .parent = parent, .rank = rank, .component_count = n };
    }

    fn deinit(self: *UnionFind, gpa: Allocator) void {
        gpa.free(self.parent);
        gpa.free(self.rank);
    }

    fn find(self: *UnionFind, x: usize) usize {
        if (self.parent[x] != x) {
            self.parent[x] = self.find(self.parent[x]); // path compression
        }
        return self.parent[x];
    }

    /// Returns true if x and y were in different components (and are now merged)
    fn unite(self: *UnionFind, x: usize, y: usize) bool {
        const root_x = self.find(x);
        const root_y = self.find(y);
        if (root_x == root_y) return false;

        // union by rank
        if (self.rank[root_x] < self.rank[root_y]) {
            self.parent[root_x] = root_y;
        } else if (self.rank[root_x] > self.rank[root_y]) {
            self.parent[root_y] = root_x;
        } else {
            self.parent[root_y] = root_x;
            self.rank[root_x] += 1;
        }
        self.component_count -= 1;
        return true;
    }

    fn sizes(self: *UnionFind, gpa: Allocator) ![]usize {
        var counts = try gpa.alloc(usize, self.parent.len);
        @memset(counts, 0);
        for (0..counts.len) |i| {
            counts[self.find(i)] += 1;
        }
        return counts;
    }
};

fn solve(comptime partId: PartId, input: []const u8, gpa: Allocator) !Result(partId) {
    const positions: []Pos = blk: {
        var lines = aoc.splitLines(aoc.trim(input));
        var positions: std.ArrayList(Pos) = .empty;
        errdefer positions.deinit(gpa);
        while (lines.next()) |line| {
            var values = aoc.split(line, ',');
            const x = try std.fmt.parseUnsigned(u32, values.next().?, 10);
            const y = try std.fmt.parseUnsigned(u32, values.next().?, 10);
            const z = try std.fmt.parseUnsigned(u32, values.next().?, 10);
            _ = try positions.append(gpa, .{ .x = x, .y = y, .z = z });
        }
        break :blk try positions.toOwnedSlice(gpa);
    };
    defer gpa.free(positions);

    var distances: DistanceQueue = blk: {
        const n = positions.len;
        const items = try gpa.alloc(DistanceInfo, n * (n - 1) / 2);
        var i: usize = 0;
        for (positions, 0..) |p1, idx1| {
            const idx2_start = idx1 + 1;
            for (positions[idx2_start..], idx2_start..) |p2, idx2| {
                const dx: u64 = @abs(@as(i64, p1.x) - @as(i64, p2.x));
                const dy: u64 = @abs(@as(i64, p1.y) - @as(i64, p2.y));
                const dz: u64 = @abs(@as(i64, p1.z) - @as(i64, p2.z));
                const dist = dx * dx + dy * dy + dz * dz;
                items[i] = .{
                    .dist = dist,
                    .junctionIdx1 = idx1,
                    .junctionIdx2 = idx2,
                };
                i += 1;
            }
        }
        break :blk DistanceQueue.fromOwnedSlice(gpa, items, {});
    };
    defer distances.deinit();

    var uf: UnionFind = try .init(gpa, positions.len);
    defer uf.deinit(gpa);

    var steps_left: u32 = switch (partId) {
        .part1 => |n| n,
        .part2 => 0,
    };

    var last_distance: ?DistanceInfo = null;

    while (distances.removeOrNull()) |d| {
        switch (partId) {
            .part1 => {
                if (steps_left == 0) break;
                steps_left -= 1;
            },
            .part2 => {},
        }

        _ = uf.unite(d.junctionIdx1, d.junctionIdx2);
        last_distance = d;

        switch (partId) {
            .part1 => {},
            .part2 => {
                if (uf.component_count == 1) break;
            },
        }
    }

    return switch (partId) {
        .part1 => blk: {
            // Count component sizes
            const sizes = try uf.sizes(gpa);
            defer gpa.free(sizes);

            var max1: usize = 0;
            var max2: usize = 0;
            var max3: usize = 0;
            for (sizes) |sz| {
                if (sz >= max1) {
                    max3 = max2;
                    max2 = max1;
                    max1 = sz;
                } else if (sz >= max2) {
                    max3 = max2;
                    max2 = sz;
                } else if (sz > max3) {
                    max3 = sz;
                }
            }
            const result: usize = max1 * max2 * max3;
            break :blk @intCast(result);
        },
        .part2 => blk: {
            const d = last_distance orelse unreachable;
            const x1: u64 = positions[d.junctionIdx1].x;
            const x2: u64 = positions[d.junctionIdx2].x;
            break :blk x1 * x2;
        },
    };
}

pub const Day = aoc.DayInfo("08", u32, u64, 97384, 9003685096, @This(), &.{.{
    .expected1 = 40,
    .expected2 = 25272,
    .input =
    \\162,817,812
    \\57,618,57
    \\906,360,560
    \\592,479,940
    \\352,342,300
    \\466,668,158
    \\542,29,236
    \\431,825,988
    \\739,650,466
    \\52,470,668
    \\216,146,977
    \\819,987,18
    \\117,168,530
    \\805,96,715
    \\346,949,466
    \\970,615,88
    \\941,993,340
    \\862,61,35
    \\984,92,344
    \\425,690,689
    ,
}});

test "samples 1" {
    const data = Day.tests[0];
    try std.testing.expectEqual(
        data.expected1,
        try solve(.{ .part1 = 10 }, data.input, std.testing.allocator),
    );
}
test "samples 2" {
    try Day.testPart2Samples();
}
test "part 1" {
    try Day.testPart1();
}
test "part 2" {
    try Day.testPart2();
}
